### ISSUE TITLE
Conway's Law

### ISSUE THREAD

**Issue by bhaugen** [valueflows/valueflows#93]
> organizations which design systems ... are constrained to produce designs which are copies of the communication structures of these organizations

> — Mel Conway

This law also works in reverse. Your system architecture will constrain the organization that can develop around it. If we want  decentralized, diverse networks of people, so must be the system architecture.

So when @fosterlynn and I set out to develop the NRP software, we knew it was the wrong architecture (a monolith), but we did not have many practical choices because of our own constraints of people, skills, money and time.

But because it was a monolith, and huge, and complex, nobody else has wanted to work on it.

We jumped into the Open Apps Ecosystem and the Value Flows as the antidote.

I write this down as an issue, because I'm seeing a lot of newer initiatives that claim to be decentralized, but in fact have hard dependencies on something that either is itself centralized (e.g. blockchains, where you have a hard dependency on one blockchain, however many copies may be out in the wild) and/or monocrops (Ethereum) where everybody needs to use one technology in order to interoperate.

Linked Open Data does not seem like that to me.

**Comment by gcassel** [comment]
^  Regarding Linked Open Data:  I know I've made at least _part_ of the following comment somewhere else recently, but I can't take time to find it right now.

Any time that metadata claims a relationship or a specific transaction between two or more agents, there is potential for conflict.   I believe we need conflict resolution tools in association with linked data.   Perhaps those tools could work as simply as a wiki, where all of the connected nodes/agents  (but no one else) had co-editing privileges.    Perhaps this only works in relation to a fractally organized reputation system?

**Comment by bhaugen** [comment]
The early stages of an actual conversation for action between two or more agents are often about aligning their vocabularies and protocols if they haven't done that before, or do not adhere to some standard. And even if they adhere to a standard, misunderstandings will happen and need to be resolved. It needs to be part of the human-level conversation.

If the same agents collaborate more often, their conversations will get smoother. This is all normal, and no technology will remove all the problems. When technologists have tried, the people did not like the results.

**Comment by gcassel** [comment]
I agree that human-level conversation is crucial, but I perceive plenty of opportunity for people with extra resources to essentially 'shape the narrative' of linked data if tools don't proactively favor consensus and collective intelligence.   Those tools/ safeguards don't necessarily have to be embedded in the linked data tool; there are many creative potentials.

I'm not trying to complicate things.  I agree that aligning vocabularies and protocols is crucial infrastructural work. Actually, I'm sure we're both busy and I'd selfishly prefer @bhaugen if you can take some time to address Arthur's and Benjamin's feedback here  https://www.hylo.com/p/11921 ....  some of it's so arcane that I can't even tell if it's over my head or not! 

(Needless to say, feel free to disagree or reply in any way to my feedback here.)

**Comment by gcassel** [comment]
^ or else maybe redirect that Hylo discussion into https://github.com/valueflows/valueflows/issues/92  which I currently need to catch up on; my bad!

**Comment by bhaugen** [comment]
I pulled Edward West's criticism over into that VF issue, and then deleted my link because I did not want to hijack Arthur and Ben's thread. I don't want to address Arthur and Benjamin's feedback, I haven't studied Ceptr enough to have even an intelligent question. But I liked Pavlik's response.

However, as I have said before, value flows does not now and I don't think should ever specify a technical substrate. It should work on anything from command lines to email to web apps to phone apps to whatever comes next. I'd be happy to try Ceptr on a project that wants to use it in the real world, for a real network. I have a lot of respect for Eric Harris-Braun's work.

**Comment by gcassel** [comment]
> value flows does not now and I don't think should ever specify a technical substrate.

Couldn't agree more.  Right now I perceive more human, flexible potential in value flows than in anything "where everybody needs to use one technology in order to interoperate".

If I comment further on the specific contrast between Ceptr and value flows, I'll probably use one of the topics I linked above.

**Comment by cristianvasquez** [comment]
Hello, 

I just love this initiative and the ongoing discussion around value-flows. 

I have some questions, as far as I understand this design aims to support decentralization, so I was wondering if there is any discussion channel on how to archive semantic interoperability in decentralized systems? ... In the Web, data interchange normally happens at many levels, using vocabularies built centrally and others built to support specific peer to peer interactions, within these interactions.

AFAIK Linked Data is essentially decentralized, but it don't impose or suggest anything related to transmitting data with shared 'meaning'.  What would be a good design to support peers agreeing to a point that they are successful in their operations? how we maintain or allow heterogeneity?

**Comment by bhaugen** [comment]
@cristianvasquez got a use case or two in mind where we could work out some details on this?

> What would be a good design to support peers agreeing to a point that they are successful in their operations?

Also: I was toying with some ideas about "networks of native VF apps interacting with one another" in one paragraph of this [wiki page about what we think VF might be used for](https://github.com/valueflows/valueflows/wiki/What-we-think-Value-Flows-might-be-used-for#two-possible-patterns-of-native-vf-apps). Please let me know if that connects at all with what you are thinking about?

**Comment by bhaugen** [comment]
@cristianvasquez p.s. how should we use [visual RDF](https://github.com/cristianvasquez/visualRDF)?

**Comment by cristianvasquez** [comment]
@bhaugen I think that 'Networks of apps interacting with one another' is more or less what I thought about. I imagine that the personal agent could keep or link some vocabulary which is personal, which in turn can be linked to other's vocabularies. (sometimes agents organize things in very different ways).

I think that a good use case would be when resource types are named and categorized differently from one community to another, such as food. I will browse the current use cases and think about something.

**Comment by cristianvasquez** [comment]
I don't find visualRDF very practical by itself... Personally, I think that a text based editor which in turn is synchronized with a visual representation would be more effective.

I've seen with interest:

https://rawgit.com/lavhal/testProj/master/VoColClient/TurtleEditor/turtle-editor.html

And 

http://vowl.visualdataweb.org/webvowl/ 

perhaps they could be linked.

**Comment by bhaugen** [comment]
@cristianvasquez 

> I will browse the current use cases and think about something.

Great!

>  I think that a text based editor which in turn is synchronized with a visual representation would be more effective.

Agreed. Want immediate interactive sync, though.

**Comment by almereyda** [comment]
The form of *design decision patterns* as presented within the paper [Cockburn, A: The Interaction of Social Issues and Software Architecture. Communications of the ACM 39(10):40-46. 1996.](http://alistair.cockburn.us/On+the+Interaction+of+Social+Issues+and+Software+Architecture) ([`doi:10.1145/236156.236165`](https://www.researchgate.net/publication/220423155_The_Interaction_of_Social_Issues_and_Software_Architecture), [preprint](https://web.archive.org/web/20060630190202/http://members.aol.com/acockburn/papers/softorg.htm)), found via http://wiki.c2.com/?ConwaysLaw, subsumises as:

> Designers must account for human social issues, not just technical concerns, when making architectural decisions.

---

[In `#ecobytes-lounge:matrix.allmende.io`](https://matrix.to/#/!jStxddszXMvUOUVHJV:matrix.allmende.io/$15089800654118GvxdW:matrix.allmende.io) I have also cited [DevOps insights into Conway's law](https://intellyx.com/2015/06/22/devops-insights-into-conways-law/):

> The causality question behind Conway’s Law, therefore, is less about how changing software organizations can lead to better software, but rather how companies can best leverage changing technology in order to transform their organizations.
> 
> Hints at how to answer this question surprisingly come from the world of devops – surprising because the focus of devops is ostensibly on building and deploying better software more quickly. Be that as it may, there’s no question that technology change is a primary facilitator and driving force for the devops cultural and organizational shifts.
> 
> ...
> 
> The end result is essentially a two-level application of Conway’s Law: a collaborative extended community of technologists that creates not simply a collection of disparate tools but rather *chainable* tools that leverage crowdsourced architectural principles to facilitate a level of coordination and interactivity we’ve never seen before.
> 
> This coordinated technology environment, in turn, facilitates the reorganization within companies, as they now have the tools they need to break down organizational silos, and people within those companies self-organize along horizontal lines

**Comment by TiberiusB** [comment]
Bob :

So when @fosterlynn <https://github.com/fosterlynn> and I set out to
develop the NRP software, we knew it was the wrong architecture (a
monolith), but we did not have many practical choices because of our own
constraints of people, skills, money and time.

Yes, there is a link between the architecture of the tech and the
organisational form, and one influences each other.  But that link is not
so hard.
I always talk about the difference between the NRP that Sensorica is still
using and the organisation itself, it's p2p culture, its governance and
methodologies.

*You can fake it!*

You can have a monolith application running on a server, on a classical
database, be conscious about it and act as if it was a p2p infrastructure.
That allows you to act as if, to consciously fake some aspects and develop
other aspects of the organisation.

Over the past 8 years Sensorica has accumulated a lot of experience in open
innovation and peer production while using the NRP. We put it under the
control of some individuals who used it in a way that made everyone else
feel secure and feel as if they were operating withing a pure p2p
environment. That allowed us to understand governance for peer production,
to build methodologies of work ans culture. At the same time, we always had
in the back of our minds the dream to one day work with a truly p2p IT
infrastructure, a p2p-rized NRP, and gathered/accumulated requirements for
such tools. Today we have some good ideas about how a p2p-rized NRP would
function,  probably better ideas that people who have never evolved within
a "fake" p2p environment like ours.

That's how things evolve in society. We can never have everything at once.
We almost always arrive at a new state in different stages, faking our way
through, until all the pieces are in place and we start to do it for real.

**Comment by bhaugen** [comment]
@TiberiusB thanks for the reflection. I totally agree with all of that. Especially this part:
> Today we have some good ideas about how a p2p-rized NRP would
function,  probably better ideas that people who have never evolved within
a "fake" p2p environment like ours.

I think, however, that a p2p architecture for the software is necessary for a good value network to emerge working on the software itself. The monolith has held us back on that. Been too difficult for new devs to get into, and carve out some piece for themselves to work on. That's starting to happen now in some of the newer projects, and eventually in the mythical [Open App Ecosystem](https://www.loomio.org/g/exAKrBUp/open-app-ecosystem).

**Comment by almereyda** [comment]
Yes, there are people who claim that even within Facebook, what is actually
happening is peer-to-peer (P2P) exchange.

It just adds up to the joke, that behind the scenes, most processes in its
infrastructure rely on distributed networking, storage and processing
protocols, all faked into a single TLD.
